README
CGI-Application-Plugin-REST

If you use the CGI::Application framework, this plugin will help you create a 
RESTful (that's the common term for "using REST") architecture by abstracting 
out a lot of the busy work needed to make it happen.

INSTALLATION

If you have the Module::Signature module installed, you can verify the
integrity of this distribution by typing:

    cpansign -v

Module::Signature requires GnuPG which you can get from: 
http://www.gnupg.org/download/
A binary version for Windows is available from that page.
        
Then to install this module, run the following commands:

    perl Build.pl
    ./Build
    ./Build test
    ./Build install
    
On Windows you don't need to type the preliminary ./

Alternatively, you can do:

    perl Makefile.PL
    make
    make test
    make install

If you are using ActiveState perl on Windows and you don't have a make 
utility, you can download nmake from:
http://download.microsoft.com/download/vc15/patch/1.52/w95/en-us/nmake15.exe
(Replace make with nmake in the instructions above if you are using it.)

If you are using Strawberry perl on Windows, it comes with a make utility 
called dmake.  (Replace make with dmake in the instructions above if you are
using it.)

DEPENDENCIES

This module is pure perl and should work with any version of perl, 5.008 or 
greater.  It requires these other modules from CPAN:

CGI::Application    4.04
REST-Utils          0.4

At build time you will need the above modules plus the following if you want
to run the tests:

Test::LongString
Test::WWW::Mechanize::CGIApp

The signature can be tested by installing the following module:

Module::Signature

and setting the environment variable TEST_SIGNATURE to 1.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc 
command.

    perldoc CGI::Application::Plugin::REST

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/CGI-Application-Plugin-REST

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=CGI-Application-Plugin-REST

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/CGI-Application-Plugin-REST

    CPAN Ratings:
        http://cpanratings.perl.org/d/CGI-Application-Plugin-REST

COPYRIGHT AND LICENCE

Copyright (C) 2010 Consolidated Braincells Inc. All rights reserved.

This distribution is free software; you can redistribute it and/or modify it
under the terms of either:

a) the GNU General Public License as published by the Free Software
Foundation; either version 2, or (at your option) any later version, or

b) the Artistic License version 2.0.

The full text of the license can be found in the LICENSE file included
with this distribution.


Changes
Revision history for CGI-Application-Plugin-REST

0.2     Fri Oct 15 07:14:00 EDT 2010
        Added a rest_resource() function to create an entire set of routes
        in one go.

0.11    Mon Oct 11 21:33:30 EDT 2010
        Added C::A::P::DevPopup to the build requirements as it is
        needed for t/devpopup.t.

0.1     Sun Oct 10 04:57:56 EDT 2010
        First version, released on an unsuspecting world.



t/lib/Test/CAPREST.pm
package Test::CAPREST;
use strict;
use warnings;
use base 'CGI::Application';
use CGI::Application::Plugin::REST qw( :all );

sub setup {
    my ($self) = @_;

    $self->run_modes([ 'default' ]);
    $self->rest_error_mode('error');
    $self->start_mode('default');

    if (defined $self->query->param('bogusdispatch')) {
        $self->rest_route( '/baz/string/*/' );
    }
    elsif (defined $self->query->param('bogusroute')) {
        $self->rest_route(
            '/zing' => [qw/ ptang krrang /],
        );
    }
    elsif (defined $self->query->param('bogusmethod')) {
        $self->rest_route(
            '/zing' => {
                'WTF' => 'ptang',
            },
        );
    }
    elsif (defined $self->query->param('bogussubroute')) {
        $self->rest_route(
            '/zing' => {
                'GET' => ['application/xml', 'arf'],
            },
        );
    }
    elsif (defined $self->query->param('routeprefix')) {
        $self->rest_route(
            '/zing' => {
                'GET' => 'zap',
            },
        );
        $self->rest_route_prefix('/app');
    }
    elsif (!defined $self->query->param('nodispatch')) {
        # Remember to change rest_route_return_value test in t/routes.t
        # when you change number of routes here. (add 1 for default '/'.)
        my $routes = {
            '/foo'                    => 'wibble',
            '/bar/:name/:id?/:email'  => \&wobble,
        };
        $self->rest_route($routes);
        $self->rest_route(
            '/baz/string/*'           => 'woop',
            '/quux'                   => {
                'GET'    => 'ptang',
                'DELETE' => 'krrang',
            },
            '/edna'                   => {
                'POST'   => 'blip',
                '*'      => 'blop',
            },
            '/grudnuk'                => {
                GET      => {
                    'application/xml' => 'zip',
                    '*/*'             => 'zap',
                },
                PUT      => {
                    'application/xml' => 'zoom',
                },
            },
            '/arf'       => {
                GET                   => 'zap',
            },
        );
    }

    return;
}

sub default {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('default') .
           $q->end_html;
}

sub error {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('error') .
           $q->end_html;
}

sub wibble {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('No parameters') .
           $q->end_html;
}

sub wobble {
    my ($self) = @_;

    my $q = $self->query;

    my $title = join q{ }, ($self->rest_param('email'),
        $self->rest_param('name'), $self->rest_param('id'));
    return $q->start_html($title) .
           $q->end_html;
}

sub woop {
    my ($self) = @_;

    my $q = $self->query;

    my $title = $self->rest_param('dispatch_uri_remainder');
    return $q->start_html($title) .
           $q->end_html;
}

sub ptang {
    my ($self) = @_;

    my $q = $self->query;

    my $title = scalar keys %{ $self->rest_route };
    return $q->start_html($title) .
           $q->end_html;
}

# krrang() intentionally omitted.

sub blip {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('blip') .
           $q->end_html;
}

sub blop {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('blop') .
           $q->end_html;
}

sub zip {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('zip') .
           $q->end_html;
}

sub zap {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('zap') .
           $q->end_html;
}

sub zoom {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('zoom') .
           $q->end_html;
}

1;


t/lib/Test/CAPRESTResource.pm
package Test::CAPRESTResource;
use strict;
use warnings;
use base 'CGI::Application';
use CGI::Application::Plugin::REST qw( :all );

sub setup {
    my ($self) = @_;

    $self->run_modes([ 'default' ]);
    $self->rest_error_mode('error');
    $self->start_mode('default');

    $self->rest_resource('widget');

    $self->rest_resource(resource => 'fidget', prefix => 'foo',
        identifier => 'num',
        in_types => ['application/xml'], out_types => ['text/html']);

    if (defined $self->query->param('noargs')) {
        $self->rest_resource();
    }

    if (defined $self->query->param('bogusargs')) {
        $self->rest_resource('midget', 'gidget', 'apt-get');
    }

    if (defined $self->query->param('bogusresource')) {
        my $resource = {
            prefix     => 'foo',
            identifier => 'num',
        };
        $self->rest_resource($resource);
    }

    if (defined $self->query->param('bogusintypes')) {
        my $resource = {
            resource     => 'midget',
            in_types     => 'foo',
        };
        $self->rest_resource($resource);
    }

    if (defined $self->query->param('bogusouttypes')) {
        my $resource = {
            resource     => 'midget',
            out_types    => 'foo',
        };
        $self->rest_resource($resource);
    }

    return;
}

sub default {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('default') .
           $q->end_html;
}

sub error {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('error') .
           $q->end_html;
}

sub widget_create {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget create') .
           $q->end_html;
}


sub widget_destroy {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget destroy ' . $self->rest_param('id')) .
           $q->end_html;
}


sub widget_edit {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget edit ' . $self->rest_param('id')) .
           $q->end_html;
}


sub widget_index {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget index') .
           $q->end_html;
}


sub widget_new {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget new') .
           $q->end_html;
}


sub widget_show {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget show ' . $self->rest_param('id')) .
           $q->end_html;
}


sub widget_update {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('widget update ' . $self->rest_param('id')) .
           $q->end_html;
}

sub foo_create {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo create') .
           $q->end_html;
}


sub foo_destroy {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo destroy ' . $self->rest_param('num')) .
           $q->end_html;
}


sub foo_edit {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo edit ' . $self->rest_param('num')) .
           $q->end_html;
}


sub foo_index {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo index') .
           $q->end_html;
}


sub foo_new {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo new') .
           $q->end_html;
}


sub foo_show {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo show ' . $self->rest_param('num')) .
           $q->end_html;
}


sub foo_update {
    my ($self) = @_;

    my $q = $self->query;

    return $q->start_html('foo update ' . $self->rest_param('num')) .
           $q->end_html;
}

1;


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    dist_name           => 'CGI-Application-Plugin-REST',
    module_name         => 'CGI::Application::Plugin::REST',
    dist_abstract       => 'Helps implement RESTful architecture in CGI applications',
    dist_author         => 'Jaldhar H. Vyas <jaldhar@braincells.com>',
    license             => 'perl',
    dist_version        => 0.11,
    #dist_version_from   => 'lib/CGI/Application/Plugin/REST.pm',
    build_requires => {
        'CGI::Application::Plugin::DevPopup' => 0,
        'Test::WWW::Mechanize::CGIApp'       => 0,
        'Test::LongString'                   => 0,
    },
    requires => {
        'perl'                          => 5.008,
        'CGI::Application'              => 4.04,
        'REST::Utils'                   => 0.4,
    },
    add_to_cleanup      => [ 'CGI-Application-Plugin-REST-*' ],
    create_makefile_pl  => 'traditional',
    sign                => 1,
    meta_merge => {
            resources => {
                repository  => 'http://github.com/jaldhar/CGI-Application-Plugin-REST',
                homepage    => 'http://jaldhar.github.com/CGI-Application-Plugin-REST',
            },
    },
);

$builder->create_build_script();
